src/number_theory/cyclotomic/rat.lean

```diff
@@ -166,3 +167,77 @@ begin
 end
 
 end is_cyclotomic_extension.rat
+
+section power_basis
+
+open is_cyclotomic_extension.rat
+
+namespace is_primitive_root
+
+/-- The integral `power_basis` of `𝓞 K` given by a primitive root of unity, where `K` is a `p ^ k`
+cyclotomic extension of `ℚ`. -/
+noncomputable def integral_power_basis [hcycl : is_cyclotomic_extension {p ^ k} ℚ K]
+  (hζ : is_primitive_root ζ ↑(p ^ k)) : power_basis ℤ (𝓞 K) :=
+let _ := is_integral_closure_adjoing_singleton_of_prime_pow hζ in by exactI
+ (adjoin.power_basis' (hζ.is_integral (p ^ k).pos)).map
+ (is_integral_closure.equiv ℤ (adjoin ℤ ({ζ} : set K)) K (𝓞 K))
```

Comment:
I find this definition hard to parse. It suggests to me that something is not setup in the right generality. E.g. something is done for a particular construction, instead of for some characteristic predicate.
But I didn't dive in to debug this further.

Reply:
I Will think about this in a week, when I will be at a computer, but don't hesitate to share your doubts about the proof. Thanks!

Reply:
Would it be better if it was instead `(ring_of_integers.equiv $ adjoin ℤ {ζ}).symm`?

Reply:
Maybe we can give a name to the isomorphism between `adjoin ℤ ({ζ} : set K)` and `𝓞 K`, then this should be a little more readable. But the `let _...` part is needed, here or somewhere else, since `is_integral_closure_adjoing_singleton_of_prime_pow` cannot be an instance.

Reply:
It should be (slightly) more readable now.